internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -411,9 +415,41 @@ func (r *BMACReconciler) reconcileAgentSpec(log logrus.FieldLogger, bmh *bmh_v1a
 	return reconcileComplete{dirty: dirty}
 }
 
+// The detached annotation is added if the BMH provisioning state is provisioned
+func (r *BMACReconciler) addBMHDetachedAnnotationIfBmhIsProvisioned(log logrus.FieldLogger, bmh *bmh_v1alpha1.BareMetalHost, agent *aiv1beta1.Agent) reconcileResult {
+
+	bmhAnnotations := bmh.ObjectMeta.GetAnnotations()
+	// Annotation already exists
+	if _, ok := bmhAnnotations[BMH_DETACHED_ANNOTATION]; ok {
+		log.Debugf("Skipping adding detached annotation. annotation already exists \n %v", agent)
+		return reconcileComplete{}
+	}
+
+	//check if we are in unbinding-pending-user-action status. If yes, we should not
+	//re-apply the detached label because ironic need to restart the host
+	boundCondition := conditionsv1.FindStatusCondition(agent.Status.Conditions, aiv1beta1.BoundCondition)
+	if boundCondition != nil && boundCondition.Reason == aiv1beta1.UnbindingPendingUserActionReason {
+		log.Debugf("Skipping adding detached annotation. pending unbound condition \n %v", agent)
+		return reconcileComplete{}
+	}
```

Comment:
There's a portion which is shared between functions ``addBMHDetachedAnnotationIfBmhIsProvisioned`` and ``addBMHDetachedAnnotationIfAgentHasStartedInstallation``, and it will be called unconditionally in the invoking function. Maybe we should consider refactoring it?